dangerzone/isolation_provider/qubes.py

```diff
@@ -0,0 +1,180 @@
+import asyncio
+import glob
+import inspect
+import io
+import logging
+import os
+import shutil
+import subprocess
+import sys
+import tempfile
+import time
+import zipfile
+from pathlib import Path
+from typing import IO, Callable, Optional
+
+from ..document import Document
+from ..util import get_resource_path
+from .base import IsolationProvider
+
+log = logging.getLogger(__name__)
+
+from ..conversion.pixels_to_pdf import PixelsToPDF
+from ..util import get_subprocess_startupinfo, get_tmp_dir
+
+CONVERTED_FILE_PATH = (
+    "/tmp/safe-output-compressed.pdf"  # FIXME won't work for parallel conversions
+)
+
+
+def read_bytes(p: subprocess.Popen, buff_size: int) -> bytes:
+    """Read bytes from stdout."""
+    return p.stdout.read(buff_size)  # type: ignore [union-attr]
+
+
+def read_int(p: subprocess.Popen) -> int:
+    """Read 2 bytes from stdout, and decode them as int."""
+    untrusted_int = p.stdout.read(2)  # type: ignore [union-attr]
+    return int.from_bytes(untrusted_int, signed=False)
+
+
+class Qubes(IsolationProvider):
+    """Uses a disposable qube for performing the conversion"""
+
+    def install(self) -> bool:
+        return True
+
+    def _convert(
+        self,
+        document: Document,
+        ocr_lang: Optional[str],
+        stdout_callback: Optional[Callable] = None,
+    ) -> bool:
+        success = False
+
+        # FIXME won't work on windows, nor with multi-conversion
+        out_dir = Path("/tmp/dangerzone")
+        if out_dir.exists():
+            shutil.rmtree(out_dir)
+        out_dir.mkdir()
+
+        # Reset hard-coded state
+        if os.path.exists(CONVERTED_FILE_PATH):
+            os.remove(CONVERTED_FILE_PATH)
+
+        percentage = 0.0
+
+        with open(document.input_filename, "rb") as f:
+            # TODO handle lack of memory to start qube
+            if getattr(sys, "dangerzone_dev", False):
+                # Use dz.ConvertDev RPC call instead, if we are in development mode.
+                # Basically, the change is that we also transfer the necessary Python
+                # code as a zipfile, before sending the doc that the user requested.
+                p = subprocess.Popen(
+                    ["/usr/bin/qrexec-client-vm", "@dispvm:dz-dvm", "dz.ConvertDev"],
+                    stdin=subprocess.PIPE,
+                    stdout=subprocess.PIPE,
```

Comment:
Important thing before we merge this that I just noticed: we need to send the `stderr`
```python
    stderr=subprocess.DEVNULL,
```

Reply:
Yes, we can change it later. The thing is, if we don't send it to DEVNULL and we end up sending it through stdout and forgetting about this, the attacker can send arbitrary data to client.

I think the Qubes folks removed control characters already, but for a time this was a possible attack vector.

Reply:
Yeap, I agree with this attack vector. Let's consume stderr properly in #430 then.

Reply:
OK. I've added it to checklist there so we don't forget.